src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Media/HostVisual.cs

```diff
@@ -382,23 +398,42 @@ private void DisconnectHostedVisualOnAllChannels()
             DUCE.Channel channel,
             bool removeChannelFromCollection)
         {
-            if (_target != null && _connectedChannels.Contains(channel))
+            Dispatcher channelDispatcher;
+            if (_target != null && _connectedChannels.TryGetValue(channel, out channelDispatcher))
             {
-                DUCE.CompositionNode.RemoveChild(
-                    _proxy.GetHandle(channel),
-                    _target._contentRoot.GetHandle(channel),
-                    channel
-                    );
-
-                //
-                // Release the targets handle. If we had duplicated the handle,
-                // then this removes the duplicated handle, otherwise just decrease
-                // the ref count for VisualTarget.
-                //
-
-                _target._contentRoot.ReleaseOnChannel(channel);
-
-                SetFlags(channel, false, VisualProxyFlags.IsContentNodeConnected);
+                // Adding commands to a channel is not thread-safe,
+                // we must do the actual work on the same dispatcher thread
+                // where the connection happened.
+                if (CoreAppContextSwitches.HostVisualDisconnectsOnWrongThread ||
```

Comment:
Adding more AppContext switches into .NET Core codebase is not ideal. Consider removing these.

Reply:
I thought I had removed this.  I must have messed up GIT somehow.  Fixed now.